android/java/org/chromium/chrome/browser/multiwindow/BraveMultiInstanceManagerApi31.java

```diff
@@ -0,0 +1,104 @@
+
+import org.chromium.base.Log;
+import org.chromium.base.supplier.ObservableSupplier;
+import org.chromium.chrome.R;
+import org.chromium.chrome.browser.app.BraveActivity;
+import org.chromium.chrome.browser.app.tabmodel.TabModelOrchestrator;
+import org.chromium.chrome.browser.app.tabmodel.TabWindowManagerSingleton;
+import org.chromium.chrome.browser.lifecycle.ActivityLifecycleDispatcher;
+import org.chromium.chrome.browser.tab.Tab;
+import org.chromium.chrome.browser.tabmodel.TabModelSelector;
+import org.chromium.chrome.browser.ui.messages.snackbar.Snackbar;
+import org.chromium.chrome.browser.ui.messages.snackbar.SnackbarManager;
+import org.chromium.chrome.browser.ui.messages.snackbar.SnackbarManager.SnackbarController;
+import org.chromium.chrome.browser.ui.messages.snackbar.SnackbarManagerProvider;
+import org.chromium.components.browser_ui.widget.MenuOrKeyboardActionController;
+import org.chromium.ui.modaldialog.ModalDialogManager;
+
+class BraveMultiInstanceManagerApi31 extends MultiInstanceManagerApi31 {
+
+    private static final String TAG = "MultiInstanceApi31";
+    private int mInstanceId;
+    private boolean mIsMoveTabsFromSettings;
+
+    BraveMultiInstanceManagerApi31(
+            Activity activity,
+            ObservableSupplier<TabModelOrchestrator> tabModelOrchestratorSupplier,
+            MultiWindowModeStateDispatcher multiWindowModeStateDispatcher,
+            ActivityLifecycleDispatcher activityLifecycleDispatcher,
+            ObservableSupplier<ModalDialogManager> modalDialogManagerSupplier,
+            MenuOrKeyboardActionController menuOrKeyboardActionController) {
+        super(
+                activity,
+                tabModelOrchestratorSupplier,
+                multiWindowModeStateDispatcher,
+                activityLifecycleDispatcher,
+                modalDialogManagerSupplier,
+                menuOrKeyboardActionController);
+    }
+
+    @Override
+    public boolean handleMenuOrKeyboardAction(int id, boolean fromMenu) {
+        if (id == org.chromium.chrome.R.id.move_to_other_window_menu_id) {
+            mIsMoveTabsFromSettings = !fromMenu;
+        }
+        return super.handleMenuOrKeyboardAction(id, fromMenu);
+    }
+
+    @Override
+    public void moveTabAction(InstanceInfo info, Tab tab, int tabAtIndex) {
+        super.moveTabAction(info, tab, tabAtIndex);
+        if (mIsMoveTabsFromSettings) {
+            mIsMoveTabsFromSettings = false;
+            TabModelSelector selector =
+                    TabWindowManagerSingleton.getInstance().getTabModelSelectorById(mInstanceId);
+            if (selector.getTotalTabCount() == 0) {
+                closeInstance(mInstanceId, INVALID_TASK_ID);
+            }
+            if (MultiWindowUtils.getInstanceCount() == 1) {
+                BraveMultiWindowUtils.updateEnableWindows(false);
+            } else {
+                try {
+                    BraveActivity braveActivity = BraveActivity.getBraveActivity();
```

Comment:
could we use Activity that we pass in the constructor? We should use `BraveActivity.getBraveActivity()` as a last resort if there is no Activity associated with a current control. You have an access to `mActivity` super's field.

Reply:
sure, I'll update it